src/content/docs/en/reference/integrations-reference.mdx

```diff
@@ -613,6 +613,121 @@ A list of all generated pages. It is an object with one property.
 
 - `pathname` - the finalized path of the page.
 
+### Defining custom hooks
+
+As an integration author, you can define custom hooks that your integration will call. This allows for runtime interactions between your integration and other integrations. Other integrations can provide implementations for your custom hooks and your integration can call those at the appropriate moment in your integration's lifecycle.
+
+#### Defining hook types
+
+Provide the type for your custom hook by augmenting Astro’s `IntegrationHooks` interface within your integration:
+
+```ts title="types.ts"
+// on an implementation file
+declare global {
+  namespace Astro {
+    interface IntegrationHooks {
+      'myLib:eventHappened': (your: string, parameters: number) => Promise<void>;
+    }
+  }
+}
+```
+
+#### Calling custom hooks
+
+Installed integrations can be accessed from the `astro:config:setup` and `astro:config:done` hooks. Loop over these to call your custom hook on each if available:
```

Comment:
I think this first sentence is too general here, and is a little unclear.  You haven't shown installing anything. Installed by who? Installed in a users project? Installed by you (somewhere) as an integration author? Is it possible that you don't need the word installed here? Would it be any less true to say "Integrations can be accessed from these hooks.."?

Also, because the example is so abstract, I'm not sure *what* is actually happening below. When this becomes a concrete example (at setup, your integration will do this thing in a user's project...) This will become more clear.

"Loop over these" grammatically is also vague in that positionally, it kind of reads like "Loop over the hooks" vs "Loop over the integrations". I'm also unclear without an example why my custom hook *needs* to access every integration (again, where? installed by the user?). Will every custom hook NEED to talk to EVERY integration installed? (Maybe it does! Would be worth an explanation.) If I've written something very custom that affects my MDX pages, it's not intuitively obvious why some other unrelated random integration I have needs this hook called. 

This is about calling your custom hook, but I can't picture what it actually might be doing here.

Reply:
> I think this first sentence is too general here, and is a little unclear. You haven't shown installing anything. Installed by who? Installed in a users project? Installed by you (somewhere) as an integration author? Is it possible that you don't need the word installed here? Would it be any less true to say "Integrations can be accessed from these hooks.."?

Installed integrations here are the integrations used on the project your integration is running.

```ts
defineConfig({
  integrations: [
    yourIntegration(),
    mdx(),
    process.env.VERCEL === '1' ? astroIcon() : null,
  ],
});
```

`yourIntegration` and `mdx` are "installed." `astroIcon` is only "installed" when running on Vercel, even though it is installed as a project dependency.

Reply:
> Is it true that you will ALWAYS want your integration to "interact" (for lack of a better word) with all of these installed integrations?

In the same sense that Astro hooks are executed for all integrations that define them, custom hooks should go through all the integrations. It is up to the integrations to implement the hook; filtering on the integration that is calling the hooks is a bit against the idea of loosely coupling things using hooks.

> You will want your custom hook called multiple times?

Using my own integration hooks as an example:
- There is [one hook](https://inox-tools.vercel.app/content-utils/git-time#itcontentgitlisted) that is called only once and only during `astro build`
- And [another hook](https://inox-tools.vercel.app/content-utils/git-time#itcontentgitresolved) that is called multiple times, twice per file, and triggers on both `astro build` and `astro dev`

The semantics of what the hook means, when it is called, and how many times it is called is up to the integration that defined it (and is calling it).

> Would an example be like, your custom hook needs to set a configuration, so you need to systematically check what each of these integrations might have set so that yours doesn't conflict, or knows what might have already been set?

I think this is more of a use case for an inter-integration API for proactive actions, not reactive (which is not a thing in Astro itself, at least not yet, but I also have a [lib](https://inox-tools.vercel.app/modular-station/api) enabling that 👉🏼👈🏼).

The use cases are much more similar to what Astro's own hooks are for. A hook just before something happens internally so integrations can modify it, another just after it happens so integrations can use the result.

Some examples from Astro's integrations:
- Sitemap could have an `astro:sitemap:discovered` hook triggered after it discovers the pages but before it generates the sitemaps to allow integrations to add or remove items programmatically (which currently can't be done)
- Sitemap could have an `astro:sitemap:done` hook triggered after it finishes generating the sitemap, which is currently unreliable since even if you use `astro:build:done`, your integration might run before the sitemap integration
- Astro DB could have a hook after the dev DB is created and seeded, giving a handle for it so integration can interact with it
- An adapter could define a hook like a `cloudflare:before:build` such that integrations can switch their internals into a Cloudflare-compatible mode with some tradeoff (like a pure JS implementation instead of a native module that would be faster)